### PR TITLE
Add new facility mode for reporting rate report

### PR DESCRIPTION
**Pull request by twymer**
screenshot 2014-12-04 12 33 01

https://docs.google.com/a/dimagi.com/document/d/1WkTqsdy4aC-gF1j0mGUCs7p_Y41jK5LGq0kxDJWF_hQ/edit

### CODE REVIEW

**Comment by czue** [comment]
assume QA passed?

**Comment by czue** [comment]
@twymer bump on question

**Comment by twymer** [comment]
@czue Wasn't formal QA, just wanted to show Ro and get feedback.

File: corehq/apps/reports/commtrack/data_sources.py [hunk 1]
```diff
@@ -418,26 +432,30 @@ def get_data(self):
             for spoint_id, loc_id in spoint_loc_map.items():
                 loc = locations[loc_id]
 
-                form_ids = StockReport.objects.filter(
+                results = StockReport.objects.filter(
                     stocktransaction__case_id=spoint_id
-                ).exclude(
-                    date__lte=self.start_date
-                ).exclude(
-                    date__gte=self.end_date
```
Comment (czue) [inline corehq/apps/reports/commtrack/data_sources.py]: why did you need to remove these again?

File: corehq/apps/reports/commtrack/data_sources.py [hunk 2]
```diff
@@ -418,26 +432,30 @@ def get_data(self):
...
                 matched = False
-                for form_id in form_ids:
+                for form_id, date in results:
+                    if self.converted_start_datetime > date:
+                        break
+
                     try:
-                        if XFormInstance.get(form_id).xmlns in form_xmlnses:
+                        if self.converted_end_datetime >= date and \
+                           XFormInstance.get(form_id).xmlns in form_xmlnses:
```
Comment (czue) [inline corehq/apps/reports/commtrack/data_sources.py]: I may have commented on this earlier or may have missed this, but it'd be much better to do this in bulk somehow if possible. hitting couch for every form is going to be really slow

Comment (czue) [inline corehq/apps/reports/commtrack/data_sources.py]: (not sure whether this is actionable since it looks like it's ~the same behavior, just noting)

**Comment by czue** [comment]
I think this is good to merge after you respond to comments, but also going to go on the record with serious performance concerns about these reports. am guessing that they'll need to be rewritten at some point but that's a larger effort that we can defer for now.

File: corehq/apps/reports/commtrack/data_sources.py (hunk 1 above)
Comment (twymer) [inline corehq/apps/reports/commtrack/data_sources.py]: To be able to get data on the last reporting date when there were no reports during the time window. The two options seemed to be this or making a second SQL query to get the value for the last, if needed.

File: corehq/apps/reports/commtrack/data_sources.py (hunk 2 above)
Comment (twymer) [inline corehq/apps/reports/commtrack/data_sources.py]: Is there a way to bulk `.get` with a list of ids?

Comment (czue) [inline corehq/apps/reports/commtrack/data_sources.py]: yeah that's basically what `iter_docs` does

File: corehq/apps/reports/commtrack/data_sources.py (hunk 1 above)
Comment (czue) [inline corehq/apps/reports/commtrack/data_sources.py]: gotcha. second sql query will probably be faster if you can push the aggregation to the database.
